language/vm/vm_runtime/src/counters.rs

```diff
@@ -16,52 +16,60 @@ const TXN_EXECUTION_DISCARD: &str = "txn.execution.discard";
+pub const TXN_VERIFICATION_TIME_TAKEN: &str = "txn_gas_verification_time_taken";
+pub const TXN_VALIDATION_TIME_TAKEN: &str = "txn_gas_validation_time_taken";
+pub const TXN_EXECUTION_TIME_TAKEN: &str = "txn_gas_execution_time_taken";
+pub const TXN_PROLOGUE_TIME_TAKEN: &str = "txn_gas_prologue_time_taken";
+pub const TXN_EPILOGUE_TIME_TAKEN: &str = "txn_gas_epilogue_time_taken";
+pub const TXN_EXECUTION_GAS_USAGE: &str = "txn_gas_execution_gas_usage";
+pub const TXN_TOTAL_GAS_USAGE: &str = "txn_gas_total_gas_usage";
 
 lazy_static::lazy_static! {
     // the main metric (move_vm)
-    static ref VM_COUNTERS: OpMetrics = OpMetrics::new_and_registered("move_vm");
+    pub static ref VM_COUNTERS: OpMetrics = OpMetrics::new_and_registered("move_vm");
 
     static ref VERIFIED_TRANSACTION: IntCounter = VM_COUNTERS.counter(TXN_VERIFICATION_SUCCESS);
     static ref BLOCK_TRANSACTION_COUNT: IntGauge = VM_COUNTERS.gauge(TXN_BLOCK_COUNT);
-
-    pub static ref TXN_TOTAL_TIME_TAKEN_HISTOGRAM: DurationHistogram = VM_COUNTERS.duration_histogram("txn.gas.total.time_taken");
-    pub static ref TXN_VERIFICATION_TIME_TAKEN_HISTOGRAM: DurationHistogram = VM_COUNTERS.duration_histogram("txn.gas.verification.time_taken");
-    pub static ref TXN_VALIDATION_TIME_TAKEN_HISTOGRAM: DurationHistogram = VM_COUNTERS.duration_histogram("txn.gas.validation.time_taken");
-    pub static ref TXN_EXECUTION_TIME_TAKEN_HISTOGRAM: DurationHistogram = VM_COUNTERS.duration_histogram("txn.gas.execution.time_taken");
-    pub static ref TXN_PROLOGUE_TIME_TAKEN_HISTOGRAM: DurationHistogram = VM_COUNTERS.duration_histogram("txn.gas.prologue.time_taken");
-    pub static ref TXN_EPILOGUE_TIME_TAKEN_HISTOGRAM: DurationHistogram = VM_COUNTERS.duration_histogram("txn.gas.epilogue.time_taken");
-    pub static ref TXN_TOTAL_GAS_USAGE_HISTOGRAM: Histogram = VM_COUNTERS.histogram("txn.gas.total.gas_usage");
-    pub static ref TXN_EXECUTION_GAS_USAGE_HISTOGRAM: Histogram = VM_COUNTERS.histogram("txn.gas.execution.gas_usage");
 }
 
+/// Wrapper around time::Instant.
 pub fn start_profile() -> Instant {
     Instant::now()
 }
 
+/// Reports the number of transactions in a block.
+pub fn report_block_count(count: usize) {
+    match i64::try_from(count) {
+        Ok(val) => BLOCK_TRANSACTION_COUNT.set(val),
+        Err(_) => BLOCK_TRANSACTION_COUNT.set(std::i64::MAX),
+    }
+}
+
 // All statistics gather operations for the time taken/gas usage should go through this macro. This
 // gives us the ability to turn these metrics on and off easily from one place.
 #[macro_export]
 macro_rules! record_stats {
-    ($e:expr) => {
-        $e
+    // Gather some information that is only needed in relation to recording statistics
+    (info | $($stmt:stmt);+;) => {
+        $($stmt);+;
     };
-    ($histrogram:expr, $block:block) => {{
+    // Per-transaction info: Set the gas histogram for $ident to be $amount.
+    (gas | $ident:ident | $amount:expr) => {
```

Comment:
IMHO we can use this macro to record all integer statistics. Should we change the name to make it sounds a bit more generic?

Reply:
Good idea. I've updated the macro to generalize this, and allow us to also set other types of statistics as well.